discovery: Add AWS EKS audit log fetching for Access Graph

camscale opened the pull request:
Extend the static config for Access Graph discovery to be able to
specify the EKS cluster for which apiserver audit logs should be fetched
and sent to Access Graph.

Extend the Access Graph discovery resource syncer to identify clusters
that should have their apiserver audit logs fetched based on discovery
configuration. Send the list of clusters over a channel for a log
fetcher to receive so it can fetch the logs and send them to access
graph.

Add a watcher to start fetchers for all access graph EKS clusters that
are configured to have Kubernetes apiserver audit logs fetched and send
them to access graph. This receives the list of clusters from the resource
discovery syncer and reconciles it against the clusters currently having
their logs fetched, starting and stopping logs fetchers as required.

Issue: https://github.com/gravitational/access-graph/issues/1589

---

Note: This PR is split into logical commits that can be reviewed separately for
easier reviewing.

camscale committed:
lib/config: Add static config for AccessGraph EKS audit logs

Extend the static config for Access Graph discovery to be able to
specify the EKS cluster for which apiserver audit logs should be fetched
and sent to Access Graph.

camscale committed:
discovery: Add AWS EKS audit log fetching for Access Graph

Add a watcher to start fetchers for all access graph EKS clusters that
are configured to have Kubernetes apiserver audit logs fetched and send
them to access graph. It receives the set of clusters to fetch audit
logs for from the AWS resource syncer as it discovers EKS clusters.
Those clusters are reconciled against the current set of log fetchers,
with no-longer-needed fetchers stopped and new fetchers started as
needed.

This commit requires go.mod be updated with:

    go get github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@latest

It is left out of this commit for now as it makes rebasing/merging
master easier.

camscale committed:
Add github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs to go.mod

Run:

    go get github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@latest
    make go-mod-tidy-all
    # Manually move the go.mod line back to the first section!?!?

This commit is kept separate for easier merging/rebasing.

juliaogris commented on lib/srv/discovery/fetchers/aws-sync/eks.go:
```diff
@@ -58,6 +60,7 @@ func (a *Fetcher) pollAWSEKSClusters(ctx context.Context, result *Resources, col
 		result.EKSClusters = output.clusters
```
pollAWSEKSClusters is a function that returns a function that fetches
 any chance that could be done with one less indirection?

juliaogris commented on lib/srv/discovery/eks_audit_log_fetcher.go:
```diff
@@ -0,0 +1,192 @@
+	"log/slog"
+	"time"
+
+	cwltypes "github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs/types"
```
cloudwatch would have made more sense to me to read.

juliaogris commented on lib/srv/discovery/access_graph_aws.go:
```diff
@@ -518,6 +518,7 @@ func (s *Server) initTAGAWSWatchers(ctx context.Context, cfg *Config) error {
 	eksAuditLogClustersCh := make(chan []eksAuditLogCluster)
 
 	if cfg.AccessGraphConfig.Enabled {
+		go s.initEKSAuditLogWatcher(ctx, eksAuditLogClustersCh)
```
i don't quite understand why this is started if it is not configured?

juliaogris commented on lib/srv/discovery/eks_audit_log_fetcher.go:
```diff
@@ -0,0 +1,192 @@
+
+// Run starts continuously polling AWS Cloud Watch Logs for Kubernetes apiserver
+// audit logs for the configured cluster, feeding the logs retrieved to the
+// configured grpc stream. It runs until the given context is canceled.
```
```
// Run continuously polls AWS CloudWatch Logs for Kubernetes apiserver audit logs
// for the configured cluster. It streams the logs to the configured gRPC stream
// and blocks until the provided context is canceled.
```

juliaogris commented on lib/srv/discovery/eks_audit_log_watcher.go:
```diff
@@ -0,0 +1,392 @@
+
+// initEKSAuditLogWatcher starts the EKS audit log watcher if there are any
+// static or dynamic configurations specifying to collect EKS audit logs.
+// If there are none, it waits for discover changes to the dynamic
```
discover changes -> discovery config changes

juliaogris commented on lib/srv/discovery/eks_audit_log_watcher.go:
```diff
@@ -0,0 +1,392 @@
+		w.fetchers[arn] = logFetcher
+		go func() {
+			err := logFetcher.Run(ctx)
+			w.completedCh <- fetcherCompleted{logFetcher, err}
```
could block forever if main run loop exits without draining completedCh. use buffered chan

camscale replied:
Good catch. Fixed.

juliaogris commented on lib/srv/discovery/eks_audit_log_fetcher.go:
```diff
@@ -0,0 +1,192 @@
+	var awsEvent cwltypes.FilteredLogEvent
+	for _, awsEvent = range awsEvents {
+		// TODO(camscale): Track event sizes and don't go over protobuf message
+		// limit. newAccessGraphClient() sets the limit to 50MB
```
Maybe add a note that FetchEKSAuditLogs limits to 500 iiuc, so unlikely to go over the 50MB

juliaogris commented on lib/srv/discovery/fetchers/aws-sync/cloudwatchlogs.go:
```diff
@@ -0,0 +1,108 @@
+
+	var result []cwltypes.FilteredLogEvent
+	for p := cloudwatchlogs.NewFilterLogEventsPaginator(client, input); p.HasMorePages(); {
+		output, err := p.NextPage(ctx)
```
check for ctx.Err() above? AI suggests it, I could not trace to wether this is really an issue

juliaogris reviewed (approved):
LGTM

juliaogris commented on lib/srv/discovery/eks_audit_log_fetcher.go:
```diff
@@ -0,0 +1,192 @@
+func (f *eksAuditLogFetcher) receiveTAGKubeAuditLogResume(ctx context.Context) (*accessgraphv1alpha.KubeAuditLogCursor, error) {
+	msg, err := f.stream.Recv()
+	if err != nil {
+		return nil, trace.Wrap(err, " failed to receive KubeAuditLogStream resume state")
```
Remove leading space.

juliaogris reviewed (approved):
Still good, just a couple of little nits.

r0mant reviewed (commented):
Can we get some test coverage going here?

camscale committed:
discovery: Refactor eks audit log fetching for testing

Refactor the eksAuditLog{Watcher,Fetcher} and the aws_sync.Fetcher
cloudwatchlogs to be more testable:

* factor away eksAuditLogFetcher from eksAuditLogWatcher. The watcher
  just needs a factory function to create a fetcher, and all the watcher
  needs from that fetcher is a `Run()` method. Lift the cancel func out
  of the watcher and store it directly in the watcher, as only the
  watcher uses it.
* factor away aws_sync.Fetcher from eksAuditLogFetcher. All it needs
  from the sync fetcher it calls is one method to fetch cloudwatch logs.
  Make that an interface and use just that. This allows a fake source of
  cloudwatch logs to be provided for testing. While here, use protobuf
  getters rather than accessing fields directly.
* Use protobuf getters in aws_sync.Fetcher cloudwatchlogs instead of
  accessing fields directly. In future, we could pass in an interface
  with those getters to make the code more testable.

camscale committed:
discovery: Add eks audit log tests

Add tests for `eksAuditLogWatcher` and `eksAuditLogFetcher`. Copy the
grpc stream testing util from the access graph repo into teleport as it
is useful for the bidirectional streaming methods uses by access graph,
and makes it easier to test on the client side.
